src/upgrades/pre_0_10_0/cas_migration.rs

```diff
@@ -0,0 +1,246 @@
+        let current_store = KeyValueStore::disk(&config.data_dir, CA_OBJECTS_DIR)?;
+        let new_store = KeyValueStore::disk(&config.upgrade_data_dir(), CA_OBJECTS_DIR)?;
+        Ok(CaObjectsMigration {
+            current_store,
+            new_store,
+        })
+    }
+
+    fn prepare_new_data_for(&self, ca: &CaHandle) -> Result<(), PrepareUpgradeError> {
+        let key = KeyStoreKey::simple(format!("{}.json", ca));
+        let old_objects: OldCaObjects = self
+            .current_store
+            .get(&key)?
+            .ok_or_else(|| PrepareUpgradeError::Custom(format!("Cannot find current objects for CA {}", ca)))?;
+
+        let converted: CaObjects = old_objects.try_into()?;
+
+        self.new_store.store(&key, &converted)?;
+
+        Ok(())
+    }
+}
+
+pub struct CasStoreMigration {
+    current_kv_store: KeyValueStore,
+    new_kv_store: KeyValueStore,
+    new_agg_store: AggregateStore<CertAuth>,
+    ca_objects_migration: CaObjectsMigration,
+}
+
+impl CasStoreMigration {
+    pub fn prepare(mode: UpgradeMode, config: &Config) -> UpgradeResult<()> {
+        let current_kv_store = KeyValueStore::disk(&config.data_dir, CASERVER_DIR)?;
+        let new_kv_store = KeyValueStore::disk(&config.upgrade_data_dir(), CASERVER_DIR)?;
+        let new_agg_store = AggregateStore::<CertAuth>::disk(&config.upgrade_data_dir(), CASERVER_DIR)?;
+        let ca_objects_migration = CaObjectsMigration::create(config)?;
+
+        CasStoreMigration {
+            current_kv_store,
+            new_kv_store,
+            new_agg_store,
+            ca_objects_migration,
+        }
+        .prepare_new_data(mode)
+    }
+}
+
+impl UpgradeStore for CasStoreMigration {
+    fn needs_migrate(&self) -> Result<bool, PrepareUpgradeError> {
+        Ok(self.current_kv_store.version_is_after(KrillVersion::release(0, 9, 0))?
+            && self
+                .current_kv_store
+                .version_is_before(KrillVersion::candidate(0, 10, 0, 1))?)
+    }
+
+    fn prepare_new_data(&self, mode: UpgradeMode) -> Result<(), PrepareUpgradeError> {
+        // check existing version, wipe if needed
+        self.preparation_store_prepare()?;
+
+        info!(
+            "Prepare upgrading CA command and event data to Krill version {}",
+            KRILL_VERSION
+        );
+
+        // Migrate the event sourced data for each CA and create new snapshots
+        for scope in self.current_kv_store.scopes()? {
+            // Getting the Handle should never fail, but if it does then we should bail out asap.
+            let handle = CaHandle::from_str(&scope)
+                .map_err(|_| PrepareUpgradeError::Custom(format!("Found invalid CA handle '{}'", scope)))?;
+
+            // Get the info from the current store to see where we are
+            let mut data_upgrade_info = self.data_upgrade_info(&scope)?;
+
+            // Migrate the initialisation event, if not done in a previous run. This
```

Comment:
In the pre_0_9_0 upgrade logic it checks if the Krill instance was setup as a test system with a Trust Anchor and bails if so, but we don't do that here. Is that because this logic is only invoked if we are already on a 0.9.0+ system and such systems cannot have a Trust Anchor at all, or is something else going on here...?

Reply:
Indeed because this update logic is only used for instances which are already on 0.9.x